hbase-server/src/main/java/org/apache/hadoop/hbase/wal/WAL.java

```diff
@@ -97,44 +97,59 @@
   void close() throws IOException;
 
   /**
-   * Append a set of edits to the WAL. The WAL is not flushed/sync'd after this transaction
-   * completes BUT on return this edit must have its region edit/sequence id assigned
-   * else it messes up our unification of mvcc and sequenceid.  On return <code>key</code> will
-   * have the region edit/sequence id filled in.
+   * Append a set of data edits to the WAL. 'Data' here means that the content in the edits will
+   * also be added to memstore.
+   * <p/>
+   * The WAL is not flushed/sync'd after this transaction completes BUT on return this edit must
+   * have its region edit/sequence id assigned else it messes up our unification of mvcc and
+   * sequenceid. On return <code>key</code> will have the region edit/sequence id filled in.
    * @param info the regioninfo associated with append
    * @param key Modified by this call; we add to it this edits region edit/sequence id.
    * @param edits Edits to append. MAY CONTAIN NO EDITS for case where we want to get an edit
-   * sequence id that is after all currently appended edits.
-   * @param inMemstore Always true except for case where we are writing a compaction completion
-   * record into the WAL; in this case the entry is just so we can finish an unfinished compaction
-   * -- it is not an edit for memstore.
+   *          sequence id that is after all currently appended edits.
    * @return Returns a 'transaction id' and <code>key</code> will have the region edit/sequence id
-   * in it.
+   *         in it.
+   * @see #appendMarker(RegionInfo, WALKeyImpl, WALEdit, boolean)
    */
-  long append(RegionInfo info, WALKeyImpl key, WALEdit edits, boolean inMemstore) throws IOException;
+  long appendData(RegionInfo info, WALKeyImpl key, WALEdit edits) throws IOException;
+
+  /**
+   * Append a marker edit to the WAL. A marker could be a FlushDescriptor, a compaction marker, or
+   * region event marker. The difference here is that, a marker will not be added to memstore.
+   * <p/>
+   * The WAL is not flushed/sync'd after this transaction completes BUT on return this edit must
+   * have its region edit/sequence id assigned else it messes up our unification of mvcc and
+   * sequenceid. On return <code>key</code> will have the region edit/sequence id filled in.
+   * @param info the regioninfo associated with append
+   * @param key Modified by this call; we add to it this edits region edit/sequence id.
+   * @param edits Edits to append. MAY CONTAIN NO EDITS for case where we want to get an edit
+   *          sequence id that is after all currently appended edits.
+   * @param closeRegion Whether this is a region close marker, i.e, the last wal edit for this
```

Comment:
Odd that  the close marker edit is an edit plus a flag set.

Reply:
We have different protobuf messages for different marker types, for example, for flush it is FlushDescriptor, for compaction it is CompactionDescriptor, for bulk load it is BulkLoadDescriptor, and for open, close region it is RegionEventDescriptor. So we have to serialize them and construct the WALEdit before calling this method, otherwise we will need to implement a lot of idfferent methods to support all these markers...

Reply:
Markers are rare. This method is rarely run. Checking for the type of the WALEdit shouldn't be that bad?

It just seems odd passing in an extra parameter specification of the Edit type when we have a mechanism for flagging special edits.

Overarching concern is that this stuff is complicated as you know and would be cool if we could avoid adding more combination types.

Not going to block the patch... just for consideration.

Reply:
I messed w/ the patch. I see how the appendData and appendMarker don't take us far enough down... down to AbstractFSWAL#appendEntry where we could ask if a close marker. But looking at patch, what would be wrong w/ doing something like this?

diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/AbstractFSWAL.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/AbstractFSWAL.java
index 2eb7c7436a..bc31204500 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/AbstractFSWAL.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/AbstractFSWAL.java
@@ -1001,7 +1001,7 @@ public abstract class AbstractFSWAL<W extends WriterBase> implements WAL {
     doAppend(writer, entry);
     assert highestUnsyncedTxid < entry.getTxid();
     highestUnsyncedTxid = entry.getTxid();
-    if (entry.isCloseRegion()) {
+    if (!entry.isInMemStore() && entry.isCloseMarker()) {
       // let's clean all the records of this region
       sequenceIdAccounting.onRegionClose(encodedRegionName);
     } else {

... where entry.isCloseMarker would do something like WALEdit.isMetaEdit...

  public boolean isMetaEdit() {
    for (Cell cell: cells) {
      if (!isMetaEditFamily(cell)) {
        return false;
      }
    }
    return true;
  }

... only instead we'd look for METAFAMILY:HBASE::CLOSE, a new define added on WALEdit?

Thanks for taking a look.

Reply:
If we did above, would we need the appendData/appendMarker distinction?

Reply:
Adding more works to the disruptor thread is a no no, it is on the write critical path, all edits will be processed by this thread. And since this is an IA.Private interface, let's use follow on issues to polish it? It can be changed even in a patched release.

What I can imagine is that, we still have the closeRegion flag in FSWALEntry, but remove it in the method parameters and in appendMarker we test whether it is a close marker. But it is still a bit annoying that, we just serialize the protobuf message before calling the method, and right in the method, we deserialize it, which really makes me unhappy...

Reply:
Sure. We'd continue testing a boolean. Rarely, we'd check the type looking for a close and we'd find a flush or compaction marker instead. Not end of the world and we could improve in follow-up.

On your this is private and we can improve in a follow on even removing parameters added by this patch, I agree.

So let me +1 this patch as is. I'll have a go at trying to do an improvement in a follow-on. Thanks.